python/cudf/cudf/core/buffer/cow_buffer.py

```diff
@@ -0,0 +1,151 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
+import copy
+from collections import defaultdict
+from typing import Any, DefaultDict, Tuple, Type, TypeVar
+from weakref import WeakSet
+
+import rmm
+
+from cudf.core.buffer.buffer import Buffer, cuda_array_interface_wrapper
+
+T = TypeVar("T", bound="CopyOnWriteBuffer")
+
+
+class CopyOnWriteBuffer(Buffer):
+    """A Buffer represents device memory.
+
+    Use the factory function `as_buffer` to create a Buffer instance.
+    """
+
+    # This dict keeps track of all instances that have the same `ptr`
+    # and `size` attributes.  Each key of the dict is a `(ptr, size)`
+    # tuple and the corresponding value is a set of weak references to
+    # instances with that `ptr` and `size`.
+    _instances: DefaultDict[Tuple, WeakSet] = defaultdict(WeakSet)
```

Comment:
This is a (slow) memory leak, I think, because when the final (ptr, size) pair for a buffer disappears it is not evicted from this dictionary.

We could either implement `__del__` on this class to cleanup the `_instances` dict when things eventually go out of scope, or make this a WeakKeyDictionary (needs a little bit of playing around):


```
class PtrSize:
    __slots__ = ("t", "__weakref__")
    def __init__(self, ptr, size):
        self.t = (ptr, size)
    def __hash__(self):
        return hash(self.t)
    def __eq__(self, o):
         return type(o) is type(self) and o.t == self.t
    def __repr__(self):
         return f"PtrSize{self.t!r}"
    def __str__(self):
        return f"PtrSize{self.t}"


class CoWBuffer:
    _instances = WeakKeyDictionary()
    ...
    def __init__(self, ...):
        self.ptrsize = PtrSize(ptr, size)
    
    def _finalize_init(self):
        self.__class__._instances.setdefault(self.ptrsize, WeakSet()).add(self)
        ...
```

Now, assuming that `ptrsize` objects are only internal, as soon as the last buffer referencing one goes away, the key disappears from the dictionary.